### ISSUE TITLE
PHP 8.1 fails tests because of Guzzle

### ISSUE THREAD

**Issue by paul-m** [GetDKAN/dkan#3898]
Search existing issues to make sure the bug has not been reported already.

## Describe the bug

This is an upstream issue with Guzzle and Drupal.

Guzzle:
- https://github.com/guzzle/guzzle/issues/3020
- https://github.com/guzzle/guzzle/issues/2971

Drupal:
- https://www.drupal.org/project/drupal/issues/3281667

DKAN 2.x branch as of commit 15a2ed0c6d0b4e5f7c94f92c72f99fc6f0d878b0.

Using DDEV to run under PHP 8.1, I see this error:
```
% ddev dkan-test-phpunit

[...]

Starting PHPUnit test run.

[..]

Runtime:       PHP 8.1.13
Configuration: /var/www/html/dkan/phpunit.xml

PHP Fatal error:  During inheritance of Countable: Uncaught Return type of GuzzleHttp\Cookie\CookieJar::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice

/var/www/html/vendor/guzzlehttp/guzzle/src/Cookie/CookieJar.php:220
/var/www/html/vendor/symfony/phpunit-bridge/DeprecationErrorHandler.php:126
/var/www/html/vendor/guzzlehttp/guzzle/src/Cookie/CookieJar.php:10
/var/www/html/vendor/fabpot/goutte/Goutte/Client.php:135
/var/www/html/vendor/symfony/browser-kit/Client.php:404
/var/www/html/vendor/friends-of-behat/mink-browserkit-driver/src/BrowserKitDriver.php:145
/var/www/html/vendor/behat/mink/src/Session.php:148
/var/www/html/vendor/weitzman/drupal-test-traits/src/GoutteTrait.php:121
/var/www/html/vendor/weitzman/drupal-test-traits/src/GoutteTrait.php:87
/var/www/html/vendor/weitzman/drupal-test-traits/src/ExistingSiteBase.php:56
/var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php:730
/var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php:665
/var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php:665
/var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php:665
/var/www/html/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:671
/var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php:148
/var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php:101
 in /var/www/html/vendor/guzzlehttp/guzzle/src/Cookie/CookieJar.php on line 10

Fatal error: During inheritance of Countable: Uncaught Return type of GuzzleHttp\Cookie\CookieJar::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice

/var/www/html/vendor/guzzlehttp/guzzle/src/Cookie/CookieJar.php:220
/var/www/html/vendor/symfony/phpunit-bridge/DeprecationErrorHandler.php:126
/var/www/html/vendor/guzzlehttp/guzzle/src/Cookie/CookieJar.php:10
/var/www/html/vendor/fabpot/goutte/Goutte/Client.php:135
/var/www/html/vendor/symfony/browser-kit/Client.php:404
/var/www/html/vendor/friends-of-behat/mink-browserkit-driver/src/BrowserKitDriver.php:145
/var/www/html/vendor/behat/mink/src/Session.php:148
/var/www/html/vendor/weitzman/drupal-test-traits/src/GoutteTrait.php:121
/var/www/html/vendor/weitzman/drupal-test-traits/src/GoutteTrait.php:87
/var/www/html/vendor/weitzman/drupal-test-traits/src/ExistingSiteBase.php:56
/var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php:730
/var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php:665
/var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php:665
/var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php:665
/var/www/html/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:671
/var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php:148
/var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php:101
 in /var/www/html/vendor/guzzlehttp/guzzle/src/Cookie/CookieJar.php on line 10
```

## Steps To Reproduce

- Set your DDEV local environment to use PHP 8.1.
- `ddev restart`
- `ddev dkan-test-phpunit`

## Expected behavior

Passing tests without errors.

**Comment by paul-m** [comment]
This is our dependency chain which leads to Guzzle ~6:
```
% ddev composer why guzzlehttp/guzzle
drupal/core               9.4.9   requires guzzlehttp/guzzle (^6.5.8 || ^7.4.5) 
drush/drush               10.6.2  requires guzzlehttp/guzzle (^6.3 || ^7.0)     
enlightn/security-checker v1.10.0 requires guzzlehttp/guzzle (^6.3|^7.0)        
fabpot/goutte             v3.3.1  requires guzzlehttp/guzzle (^6.0)             
getdkan/dkan              2.x-dev requires guzzlehttp/guzzle (^6.3)             
getdkan/harvest           1.0.1   requires guzzlehttp/guzzle (^6.3)  
```
We could change the requirement for getdkan/dkan (and harvest by extension), but that leaves us with fabpot/goutte.

Goutte is required by weitzman/drupal-test-traits, which has fixed it here: https://gitlab.com/weitzman/drupal-test-traits/-/issues/80 However, that's the 2.x dev branch and does not have a release.

**Comment by jastraat** [comment]
Looks like there are some new tags in the last month for weitzman/drupal-test-traits:  https://gitlab.com/weitzman/drupal-test-traits/-/tags

**Comment by jastraat** [comment]
I was able to get all automated tests to pass locally with PHP 8.1 by updating weitzman/drupal-test-traits to 2.0.1, adjusting other related guzzlehttp dependencies, and making a number of small changes to DKAN. I'll post all of these a PR shortly.

auto_detect_line_endings is deprecated and this call to it: https://github.com/GetDKAN/dkan/blob/2.x/modules/datastore/modules/datastore_mysql_import/src/Service/MysqlImport.php#L94 makes the related MysqlImportTest.php fail. My solution was removing the lines adding and then removing that deprecated setting.
https://php.watch/versions/8.1/auto_detect_line_endings-ini-deprecated

In addition, the Drupal core recommended package still requires version 6. To temporarily get around this, I patched guzzlehttp/guzzle in a few places. (Adding #[\ReturnTypeWillChange] where needed)

There are also issues in the following related packages:

- https://github.com/fmizzell/maquina
  PHP Fatal error:  During inheritance of JsonSerializable: Uncaught Return type of Maquina\StateMachine\Machine::jsonSerialize()
- https://github.com/m1x0n/opis-error-presenter
  PHP Fatal error:  During inheritance of JsonSerializable: Uncaught Return type of PresentedValidationError::jsonSerialize()
- https://github.com/GetDKAN/csv-parser
  PHP Fatal error:  During inheritance of JsonSerializable: Uncaught Return type of CsvParser\Parser\Csv::jsonSerialize() 
- https://github.com/GetDKAN/procrastinator
  PHP Fatal error:  During inheritance of JsonSerializable: Uncaught Return type of Result::jsonSerialize() and Job::jsonSerialize() 

Almost all the above packages can be made compatible by add #[\ReturnTypeWillChange] in a comment above the functions with incompatible return types.

**Comment by paul-m** [comment]
This is fixed as of DKAN release 2.15.
